## Part A 1
The Operation, AC_Char and city_pair datasets are properly formatted in csv format by simply removing  irrelevant values which can be checked in folders.


We proceed the problem by following thses steps
1. import "Operation" dataset
2. remove irrelevant columns
3. Sum up total hours for each rows and put it in "Total" Column
4. Group same Aircraft type and add their Total hours of flight
5. Import AC_Char dataset which has cost per flight hour column
6. Merge both Operation and AC_char dataset on same Aircraft type(inner join)
7. Convert CostPerHour in proper format(int type)
8. Multiply Total hours and Cost Per Hour

In [229]:
import numpy as np
import pandas as pd

In [230]:
operation=pd.read_csv("Operation.csv",index_col="Aircraft Type",)

In [231]:
operation.head()

Aircraft Name  jan  feb  march  april  may  june  july  august  \
Aircraft Type                                                                  
A320           A             250  251  248    253    243  245   248   246      
A330           B             300  301  298    303    293  295   298   296      
B737           C             245  246  243    248    238  240   243   241      
B747           D             310  311  308    313    303  305   308   306      
Q400           E             200  201  198    203    193  195   198   196      

               sept  oct  nov  dec  Unnamed: 14  Unnamed: 15  Unnamed: 16  \
Aircraft Type                                                               
A320           256   256  251  252 NaN          NaN          NaN            
A330           306   306  301  302 NaN          NaN          NaN            
B737           251   251  246  247 NaN          NaN          NaN            
B747           316   316  311  312 NaN          NaN          NaN            
Q400           206   206  201  202 NaN          NaN          NaN            

               Unnamed: 17  
Aircraft Type               
A320          NaN           
A330          NaN           
B737          NaN           
B747          NaN           
Q400          NaN

We have to remove last 4 Unnamed Columns and "Aircraft Name" Column

In [232]:
operation.drop('Aircraft Name',axis=1,inplace=True)

In [233]:
operation.dropna(axis=1,inplace=True) #since last 4 unnamed columns has Nan Values inside it, So we remove all columns with NaN value

## Step 3:-Sum up total hours for each rows and put it in "Total" Column

In [234]:
operation["Total"]=operation.sum(axis=1)

In [235]:
operation.head()

jan  feb  march  april  may  june  july  august  sept  oct  \
Aircraft Type                                                               
A320           250  251  248    253    243  245   248   246     256   256   
A330           300  301  298    303    293  295   298   296     306   306   
B737           245  246  243    248    238  240   243   241     251   251   
B747           310  311  308    313    303  305   308   306     316   316   
Q400           200  201  198    203    193  195   198   196     206   206   

               nov  dec  Total  
Aircraft Type                   
A320           251  252  2999   
A330           301  302  3599   
B737           246  247  2939   
B747           311  312  3719   
Q400           201  202  2399

## Step 4 :-Group same Aircraft type and add their Total hours of flight

In [236]:
operation=operation.sum(level="Aircraft Type")

In [237]:
operation

jan   feb  march  april   may  june  july  august  sept   oct  \
Aircraft Type                                                                   
A320           1240  1245  1230   1255   1205  1215  1230  1220    1270  1270   
A330           1820  1826  1808   1838   1778  1790  1808  1796    1856  1856   
B737           985   989   977    997    957   965   977   969     1009  1009   
B747           920   923   914    929    899   905   914   908     938   938    
Q400           790   794   782    802    762   770   782   774     814   814    
ATR72          795   799   787    807    767   775   787   779     819   819    

                nov   dec  Total  
Aircraft Type                     
A320           1245  1250  14875  
A330           1826  1832  21834  
B737           989   993   11816  
B747           923   926   11037  
Q400           794   798   9476   
ATR72          799   803   9536

## Step 5 import AC_Char dataset

In [238]:
ac_char=pd.read_csv("AC_Char.csv",index_col="Aircraft Type", )

In [239]:
ac_char.head()

Range (Km)  Ave. Speed (km/h)  Number of Seats  \
Aircraft Type                                                  
A320            5,000     800                150               
A330            8,000     900                250               
B737            5,000     800                150               
B747            10,000    900                350               
Q400            1,500     750                90                

              Costs per flight hour  
Aircraft Type                        
A320            5,000                
A330            7,500                
B737            5,100                
B747            12,000               
Q400            3,500

## Step6:- Inner join on both datasets

In [240]:
Combined=pd.merge(operation, ac_char, on='Aircraft Type')## By default it does inner join 

In [241]:
Combined.rename(columns={"Costs per flight hour":"costph"},inplace=True)

In [242]:
Combined

jan   feb  march  april   may  june  july  august  sept   oct  \
Aircraft Type                                                                   
A320           1240  1245  1230   1255   1205  1215  1230  1220    1270  1270   
A330           1820  1826  1808   1838   1778  1790  1808  1796    1856  1856   
B737           985   989   977    997    957   965   977   969     1009  1009   
B747           920   923   914    929    899   905   914   908     938   938    
Q400           790   794   782    802    762   770   782   774     814   814    
ATR72          795   799   787    807    767   775   787   779     819   819    

                nov   dec  Total Range (Km)  Ave. Speed (km/h)  \
Aircraft Type                                                    
A320           1245  1250  14875   5,000     800                 
A330           1826  1832  21834   8,000     900                 
B737           989   993   11816   5,000     800                 
B747           923   926   11037   10,000    900                 
Q400           794   798   9476    1,500     750                 
ATR72          799   803   9536    1,000     650                 

               Number of Seats    costph  
Aircraft Type                             
A320           150               5,000    
A330           250               7,500    
B737           150               5,100    
B747           350               12,000   
Q400           90                3,500    
ATR72          75                2,750

Drop columns which are not necessary right now...like Range, Ave Speed, Number of seats

In [243]:
Combined.drop(["Range (Km)",
"Ave. Speed (km/h)",
"Number of Seats"],axis=1,inplace=True)

In [244]:
Combined

jan   feb  march  april   may  june  july  august  sept   oct  \
Aircraft Type                                                                   
A320           1240  1245  1230   1255   1205  1215  1230  1220    1270  1270   
A330           1820  1826  1808   1838   1778  1790  1808  1796    1856  1856   
B737           985   989   977    997    957   965   977   969     1009  1009   
B747           920   923   914    929    899   905   914   908     938   938    
Q400           790   794   782    802    762   770   782   774     814   814    
ATR72          795   799   787    807    767   775   787   779     819   819    

                nov   dec  Total    costph  
Aircraft Type                               
A320           1245  1250  14875   5,000    
A330           1826  1832  21834   7,500    
B737           989   993   11816   5,100    
B747           923   926   11037   12,000   
Q400           794   798   9476    3,500    
ATR72          799   803   9536    2,750

## Step7:- Change Costph to  int format

In [245]:
Combined["costph"]=Combined["costph"].str.replace(',', '').astype(int)

## Step 8:-Multiplying Total hours and cost per hour

In [246]:
Combined["final_cost"]=Combined["Total"]*Combined["costph"]

## Part A 1 Solution

In [247]:
Combined["final_cost"]

Aircraft Type
A320     74375000 
A330     163755000
B737     60261600 
B747     132444000
Q400     33166000 
ATR72    26224000 
Name: final_cost, dtype: int64

In [248]:
Combined.rename(columns={"Total":"Totalhours"},inplace=True)

In [249]:
Combined

jan   feb  march  april   may  june  july  august  sept   oct  \
Aircraft Type                                                                   
A320           1240  1245  1230   1255   1205  1215  1230  1220    1270  1270   
A330           1820  1826  1808   1838   1778  1790  1808  1796    1856  1856   
B737           985   989   977    997    957   965   977   969     1009  1009   
B747           920   923   914    929    899   905   914   908     938   938    
Q400           790   794   782    802    762   770   782   774     814   814    
ATR72          795   799   787    807    767   775   787   779     819   819    

                nov   dec  Totalhours  costph  final_cost  
Aircraft Type                                              
A320           1245  1250  14875       5000    74375000    
A330           1826  1832  21834       7500    163755000   
B737           989   993   11816       5100    60261600    
B747           923   926   11037       12000   132444000   
Q400           794   798   9476        3500    33166000    
ATR72          799   803   9536        2750    26224000

## Part A2

Steps involved
1.  Merge combined and ac_char dataset on Airctraft Type (inner join)

2.  Find total distance by multiplying total hours and average speed of each flight
3.  Find total cost per seat per km

## Step1:-Merge combined and ac_char dataset on Airctraft Type (inner join)

In [250]:
Combined=pd.merge(Combined, ac_char, on='Aircraft Type')

In [251]:
Combined

jan   feb  march  april   may  june  july  august  sept   oct  \
Aircraft Type                                                                   
A320           1240  1245  1230   1255   1205  1215  1230  1220    1270  1270   
A330           1820  1826  1808   1838   1778  1790  1808  1796    1856  1856   
B737           985   989   977    997    957   965   977   969     1009  1009   
B747           920   923   914    929    899   905   914   908     938   938    
Q400           790   794   782    802    762   770   782   774     814   814    
ATR72          795   799   787    807    767   775   787   779     819   819    

                nov   dec  Totalhours  costph  final_cost Range (Km)  \
Aircraft Type                                                          
A320           1245  1250  14875       5000    74375000     5,000      
A330           1826  1832  21834       7500    163755000    8,000      
B737           989   993   11816       5100    60261600     5,000      
B747           923   926   11037       12000   132444000    10,000     
Q400           794   798   9476        3500    33166000     1,500      
ATR72          799   803   9536        2750    26224000     1,000      

               Ave. Speed (km/h)  Number of Seats Costs per flight hour  
Aircraft Type                                                            
A320           800                150               5,000                
A330           900                250               7,500                
B737           800                150               5,100                
B747           900                350               12,000               
Q400           750                90                3,500                
ATR72          650                75                2,750

## Step2:-Find total distance by multiplying total hours and average speed of each flight

In [252]:
Combined["TotalDistance"]=Combined["Totalhours"]*Combined["Ave. Speed (km/h)"]

In [253]:
Combined

jan   feb  march  april   may  june  july  august  sept   oct  \
Aircraft Type                                                                   
A320           1240  1245  1230   1255   1205  1215  1230  1220    1270  1270   
A330           1820  1826  1808   1838   1778  1790  1808  1796    1856  1856   
B737           985   989   977    997    957   965   977   969     1009  1009   
B747           920   923   914    929    899   905   914   908     938   938    
Q400           790   794   782    802    762   770   782   774     814   814    
ATR72          795   799   787    807    767   775   787   779     819   819    

                nov   dec  Totalhours  costph  final_cost Range (Km)  \
Aircraft Type                                                          
A320           1245  1250  14875       5000    74375000     5,000      
A330           1826  1832  21834       7500    163755000    8,000      
B737           989   993   11816       5100    60261600     5,000      
B747           923   926   11037       12000   132444000    10,000     
Q400           794   798   9476        3500    33166000     1,500      
ATR72          799   803   9536        2750    26224000     1,000      

               Ave. Speed (km/h)  Number of Seats Costs per flight hour  \
Aircraft Type                                                             
A320           800                150               5,000                 
A330           900                250               7,500                 
B737           800                150               5,100                 
B747           900                350               12,000                
Q400           750                90                3,500                 
ATR72          650                75                2,750                 

               TotalDistance  
Aircraft Type                 
A320           11900000       
A330           19650600       
B737           9452800        
B747           9933300        
Q400           7107000        
ATR72          6198400

In [254]:
Combined["AnsOfPartA2"]=(Combined["final_cost"]/Combined["Number of Seats"])/Combined["TotalDistance"]

In [255]:
Combined

jan   feb  march  april   may  june  july  august  sept   oct  \
Aircraft Type                                                                   
A320           1240  1245  1230   1255   1205  1215  1230  1220    1270  1270   
A330           1820  1826  1808   1838   1778  1790  1808  1796    1856  1856   
B737           985   989   977    997    957   965   977   969     1009  1009   
B747           920   923   914    929    899   905   914   908     938   938    
Q400           790   794   782    802    762   770   782   774     814   814    
ATR72          795   799   787    807    767   775   787   779     819   819    

               ...   dec  Totalhours  costph  final_cost  Range (Km)  \
Aircraft Type  ...                                                     
A320           ...  1250  14875       5000    74375000     5,000       
A330           ...  1832  21834       7500    163755000    8,000       
B737           ...  993   11816       5100    60261600     5,000       
B747           ...  926   11037       12000   132444000    10,000      
Q400           ...  798   9476        3500    33166000     1,500       
ATR72          ...  803   9536        2750    26224000     1,000       

              Ave. Speed (km/h)  Number of Seats  Costs per flight hour  \
Aircraft Type                                                             
A320           800               150               5,000                  
A330           900               250               7,500                  
B737           800               150               5,100                  
B747           900               350               12,000                 
Q400           750               90                3,500                  
ATR72          650               75                2,750                  

              TotalDistance  AnsOfPartA2  
Aircraft Type                             
A320           11900000      0.041667     
A330           19650600      0.033333     
B737           9452800       0.042500     
B747           9933300       0.038095     
Q400           7107000       0.051852     
ATR72          6198400       0.056410     

[6 rows x 21 columns]

In [256]:
Combined["AnsOfPartA2"]

Aircraft Type
A320     0.041667
A330     0.033333
B737     0.042500
B747     0.038095
Q400     0.051852
ATR72    0.056410
Name: AnsOfPartA2, dtype: float64

## Ans:- A330 Is Lowest

## PartB 1

Steps involved
1.  Import city pair datasets
2.  Edit ac_char dataset and change column to type int
3.  Create a function to pick only those flights which has sufficient range
4.  Find out total no. of times a flight will have to do updown to meet the demand
    ---->A320 has capacity of 150 seats and range of 5000km
    ---->AA to BB we have demand of 420 and distance of 3000km
    ---->Clearly the flight has sufficient range..Now let's talk about capacity
    ---->Total passengers=420
    ---->>In one trip A320 will take 150 passengers from AA to BB (Total trip=1)
    ---->Flight will come back(Total trip=2)
    ---->In second trip will take anpother 150 passengers from AA to BB (Total trip=3)
    ---->Flight will come back(Total trip=4)
    ---->In third trip flight will take 120 passengers from AA to BB(Total trip=5)
    ---->In 5 updown flight A320 will meet the demand
5.  Pick only those flights which meet the demand in 3 or less trips
6.  Out the the flights in 5th Steps, pick the one with minimum price

In [257]:
citypair=pd.read_csv("CityPair.csv")


In [258]:
citypair["Distance"]=citypair["Distance "].str.replace(',', '').astype(int)

In [259]:
citypair

Origin  Desitnation  Demand Distance   Distance
0  AA       BB         420      3,000    3000    
1  BB       CC         450      6,500    6500    
2  CC       AA         300      400      400     
3  AA       DD         300      1,000    1000

In [260]:
ac_char["Costs per flight hour"]=ac_char["Costs per flight hour"].str.replace(',', '').astype(int)

In [261]:
ac_char["Range (Km)"]=ac_char["Range (Km)"].str.replace(',', '').astype(int)

In [262]:
ac_char

Range (Km)  Ave. Speed (km/h)  Number of Seats  \
Aircraft Type                                                   
A320           5000        800                150               
A330           8000        900                250               
B737           5000        800                150               
B747           10000       900                350               
Q400           1500        750                90                
ATR72          1000        650                75                

               Costs per flight hour  
Aircraft Type                         
A320           5000                   
A330           7500                   
B737           5100                   
B747           12000                  
Q400           3500                   
ATR72          2750

In [263]:
def func(value):
    a=list()
    for i,j in ac_char["Range (Km)"].iteritems():
        if j>value:
            a.append(i)
    return(a)

In [264]:
size=len(citypair)

In [265]:
citypair["AircraftSL"]=np.nan

In [266]:
for i in range(size):
    citypair["AircraftSL"][i]=func(citypair["Distance"][i])

C:\Users\Prasoon\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
C:\Users\Prasoon\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\core\indexing.py:190: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)


## These are the aircrafts which can fulfill the range requirement for each city pair

In [267]:
 citypair["AircraftSL"]

0    [A320, A330, B737, B747]             
1    [A330, B747]                         
2    [A320, A330, B737, B747, Q400, ATR72]
3    [A320, A330, B737, B747, Q400]       
Name: AircraftSL, dtype: object

In [268]:
citypair

Origin  Desitnation  Demand Distance   Distance  \
0  AA       BB         420      3,000    3000       
1  BB       CC         450      6,500    6500       
2  CC       AA         300      400      400        
3  AA       DD         300      1,000    1000       

                              AircraftSL  
0  [A320, A330, B737, B747]               
1  [A330, B747]                           
2  [A320, A330, B737, B747, Q400, ATR72]  
3  [A320, A330, B737, B747, Q400]

In [269]:
citypair.drop("Distance ",axis=1,inplace=True)#Remove one Distance

In [270]:
citypair

Origin  Desitnation  Demand  Distance                             AircraftSL
0  AA       BB         420     3000      [A320, A330, B737, B747]             
1  BB       CC         450     6500      [A330, B747]                         
2  CC       AA         300     400       [A320, A330, B737, B747, Q400, ATR72]
3  AA       DD         300     1000      [A320, A330, B737, B747, Q400]

In [271]:
import math

In [272]:
def demand(value, aircraft):
    a=dict()
    for i,j in ac_char["Number of Seats"].iteritems():
        if i in aircraft:
            a[i]=(math.ceil(value/j))*2-1
    return(a)

Create a column of empty dictionary to keep the note of no. of trips each of the flight will take for each pair of cities

In [273]:
citypair["ShortlistBySize"]= [{} for _ in range(len(citypair))]

In [274]:
pd.set_option('display.max_colwidth', -1)# See complete rows columns without truncating

In [275]:
for i in range(size):
    citypair["ShortlistBySize"][i]=demand(citypair["Demand"][i],citypair["AircraftSL"][i])

C:\Users\Prasoon\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [276]:
citypair["ShortlistBySize"]

0    {'A320': 5, 'A330': 3, 'B737': 5, 'B747': 3}                       
1    {'A330': 3, 'B747': 3}                                             
2    {'A320': 3, 'A330': 3, 'B737': 3, 'B747': 1, 'Q400': 7, 'ATR72': 7}
3    {'A320': 3, 'A330': 3, 'B737': 3, 'B747': 1, 'Q400': 7}            
Name: ShortlistBySize, dtype: object

In [277]:
citypair.rename(columns={"ShortlistBySize":"TotalUpDowntrips"},inplace=True)

In [278]:
citypair

Origin  Desitnation  Demand  Distance  \
0  AA       BB         420     3000       
1  BB       CC         450     6500       
2  CC       AA         300     400        
3  AA       DD         300     1000       

                              AircraftSL  \
0  [A320, A330, B737, B747]                
1  [A330, B747]                            
2  [A320, A330, B737, B747, Q400, ATR72]   
3  [A320, A330, B737, B747, Q400]          

                                                      TotalUpDowntrips  
0  {'A320': 5, 'A330': 3, 'B737': 5, 'B747': 3}                         
1  {'A330': 3, 'B747': 3}                                               
2  {'A320': 3, 'A330': 3, 'B737': 3, 'B747': 1, 'Q400': 7, 'ATR72': 7}  
3  {'A320': 3, 'A330': 3, 'B737': 3, 'B747': 1, 'Q400': 7}

In [279]:
def price(updown):
    a=dict()
    for i,j in ac_char["Costs per flight hour"].iteritems():
        if i in updown and updown[i]<=3:
            a[i]=j
    b=min(a, key=a.get)
    return(a,b)

In [280]:
citypair["Cost"]=[{} for _ in range(len(citypair))] 

In [281]:
citypair["MinimumCostFlight"]= np.nan #To figure out which flight has minimum cost

Cost of those flights which meet the demand in 3 or less trips

In [282]:
for i in range(size):
    citypair["Cost"][i],citypair["MinimumCostFlight"][i]=price(citypair["TotalUpDowntrips"][i])

C:\Users\Prasoon\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [284]:
citypair

Origin  Desitnation  Demand  Distance  \
0  AA       BB         420     3000       
1  BB       CC         450     6500       
2  CC       AA         300     400        
3  AA       DD         300     1000       

                              AircraftSL  \
0  [A320, A330, B737, B747]                
1  [A330, B747]                            
2  [A320, A330, B737, B747, Q400, ATR72]   
3  [A320, A330, B737, B747, Q400]          

                                                      TotalUpDowntrips  \
0  {'A320': 5, 'A330': 3, 'B737': 5, 'B747': 3}                          
1  {'A330': 3, 'B747': 3}                                                
2  {'A320': 3, 'A330': 3, 'B737': 3, 'B747': 1, 'Q400': 7, 'ATR72': 7}   
3  {'A320': 3, 'A330': 3, 'B737': 3, 'B747': 1, 'Q400': 7}               

                                                        Cost MinimumCostFlight  
0  {'A330': 7500, 'B747': 12000}                              A330              
1  {'A330': 7500, 'B747': 12000}                              A330              
2  {'A320': 5000, 'A330': 7500, 'B737': 5100, 'B747': 12000}  A320              
3  {'A320': 5000, 'A330': 7500, 'B737': 5100, 'B747': 12000}  A320

## Clearly Minimum Cost Flight gives the Flight which has sufficient range, can manage the demands in 3 or less updown trips  and has minimum costs`

## So A330 and A220 are best suitable options